### PR TITLE
Feature/fgkf 601

### PR DESCRIPTION
**Pull request by MatthewWaanders**

### CODE REVIEW

File: .homeycompose/drivers/settings/buttonSequence.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "type": "text",
+  "label": {
+    "en": "Sequence",
+    "nl": "Reeks"
+  },
+  "value": "0",
+  "pattern": "^0$|^[1-6];([1-6];)?([1-6];)?([1-6];)?[1-6];?",
+  "hint": {
+    "en": "Separate button numbers by a semicolon ( ; ), use 2 to 5 buttons. No duplicate sequences are allowed. 0 = sequence disabled. Buttons are numbered from left to right, from top to bottom (1 - 6).",
```
Comment (RobinBol) [inline .homeycompose/drivers/settings/buttonSequence.json]: No {{zwaveParameterIndex}} needed?

File: drivers/FGKF-601/device.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        });
+        this.registerSetting('sequence_5', (newValue) => {
+            return this.sequenceParser(newValue);
+        });
+        this.registerSetting('sequence_6', (newValue) => {
+            return this.sequenceParser(newValue);
+        });
+
+        if (this.node && typeof this.node.CommandClass.COMMAND_CLASS_CENTRAL_SCENE !== 'undefined') {
+            this.node.CommandClass.COMMAND_CLASS_CENTRAL_SCENE.on('report', (command, report) => {
```
Comment (RobinBol) [inline drivers/FGKF-601/device.js]: Better to use `registerReportListener()` for this (a zwave-device method).

File: drivers/FGKF-601/driver.compose.json [hunk 3]
```diff
@@ -0,0 +1,124 @@
+{
+  "id": "FGKF-601",
+  "name": {
+    "en": "Keyfob"
+  },
+  "class": "other",
+  "capabilities": [],
```
Comment (RobinBol) [inline drivers/FGKF-601/driver.compose.json]: Not even battery capability?

File: drivers/FGKF-601/driver.compose.json [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      4097
+    ],
+    "productId": [
+      4096
+    ],
+    "zwaveAllianceProductId": 2064,
+    "zwaveAllianceProductDocumentation": "https://Products.Z-WaveAlliance.org/ProductManual/File?folder=&filename=Manuals/2064/FGKF-601-EN-T-v1.0_30.11.2016.pdf",
+    "learnmode": {
+      "instruction": {
+        "en": "Press any button on the Fibaro Keyfob 3 times.",
```
Comment (RobinBol) [inline drivers/FGKF-601/driver.compose.json]: {{driverNameEn}}

File: drivers/FGKF-601/driver.compose.json [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        "nl": "Druk driemaal op een willekeurige knop op de Fibaro Keyfob."
+      }
+    },
+    "unlearnmode": {
+      "instruction": {
+        "en": "Press O and – shortly at the same time, then press Δ or X shortly until the LED is green, finally press + to confirm the removal.",
+        "nl": "Druk O en – tegelijkertijd kort in, druk vervolgens kort op Δ of X tot de LED groen is, druk als laatste stap + kort in om het verwijderen te bevestigen."
+      }
+    },
+    "associationGroups": [
```
Comment (RobinBol) [inline drivers/FGKF-601/driver.compose.json]: Are all these really necessary? Usually only the first is needed for the controller.

File: drivers/FGKF-601/driver.settings.compose.json [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+      "nl": "Associatie Groep Knop Waarden"
+    },
+    "collapsed": true,
+    "children": [
+      {
+        "id": "group_operating_mode",
+        "type": "dropdown",
+        "label": {
+          "en": "Association groups operating mode",
+          "nl": "Associatie groepen werkingsmodus"
```
Comment (RobinBol) [inline drivers/FGKF-601/driver.settings.compose.json]: 'werkingsmodus' is a bit of a weird translation:p

File: drivers/FGKF-601/driver.settings.compose.json [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+          "index": 10,
+          "size": 1
+        },
+        "value": "0",
+        "values": [
+          {
+            "id": "0",
+            "label": {
+              "en": "Single press: toggle state.",
+              "nl": "Eenmaal drukken: schakel status."
```
Comment (RobinBol) [inline drivers/FGKF-601/driver.settings.compose.json]: This setting uses a dot at the end of the label (for the dropdown values), I don't think that is how you consistently use it?

File: drivers/FGKF-601/driver.settings.compose.json [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+          {
+            "id": "2",
+            "label": {
+              "en": "Single press: turn on, double press: turn off.",
+              "nl": "Eenmaal drukken: schakel aan, tweemaal drukken: sckakel uit."
+            }
+          }
+        ],
+        "hint": {
+          "en": "This parameter ({{zwaveParameterIndex}}) determines the behaviour of the KeyFob with regard to its association groups.",
```
Comment (RobinBol) [inline drivers/FGKF-601/driver.settings.compose.json]: driverNameEn

File: drivers/FGKF-601/driver.compose.json (hunk 5 above)
Comment (MatthewWaanders) [reply drivers/FGKF-601/driver.compose.json]: Yeah these are all the multilevel associations for the separate buttons

File: drivers/FGKF-601/driver.compose.json [hunk 9]
```diff
@@ -21,8 +23,8 @@
     "zwaveAllianceProductDocumentation": "https://Products.Z-WaveAlliance.org/ProductManual/File?folder=&filename=Manuals/2064/FGKF-601-EN-T-v1.0_30.11.2016.pdf",
     "learnmode": {
       "instruction": {
-        "en": "Press any button on the Fibaro Keyfob 3 times.",
-        "nl": "Druk driemaal op een willekeurige knop op de Fibaro Keyfob."
+        "en": "Press any button on the {{driverNameEN}} 3 times.",
```
Comment (RobinBol) [inline drivers/FGKF-601/driver.compose.json]: This does not work, it is `driverNameEn` and `driverNameNl`, used multiple times in this commit, not going to mark them all.

File: app.json [hunk 10]
```diff
@@ -2701,7 +2701,7 @@
               },
               "zwave": {
                 "index": 21,
-                "size": 1
+                "size": 0
```
Comment (RobinBol) [inline app.json]: Why are there a lot of `size: 0` key-value pairs?

File: drivers/FGKF-601/device.js [hunk 11]
```diff
@@ -13,7 +13,11 @@ class FibaroKeyfob extends ZwaveDevice {
         this._sequenceFlowTrigger = new Homey.FlowCardTriggerDevice('FGKF-601-sequence').registerRunListener(this._sequenceRunListener).register();
 
         // Parsing of sequences before sending to Keyfob
+        this.registerSetting('lock_timeout', (newValue) => {
+            this._enableLockMode({lock_timeout: newValue});
```
Comment (RobinBol) [inline drivers/FGKF-601/device.js]: This does not return, is that on purpose?

File: drivers/FGKF-601/device.js [hunk 12]
```diff
@@ -59,6 +63,42 @@ class FibaroKeyfob extends ZwaveDevice {
         }
     }
 
+    _enableLockMode(newValueObj) {
+	    if ( (typeof newValueObj.lock_timeout === 'number' &&
+                newValueObj.lock_timeout > 0 &&
+                this.getSetting('sequence_lock')) ||
+            (typeof newValueObj.sequence_lock === 'string' &&
+                this.getSetting('lock_timeout') > 0)) {
+	                this.node.CommandClass.COMMAND_CLASS_PROTECTION.PROTECTION_SET({
```
Comment (RobinBol) [inline drivers/FGKF-601/device.js]: Use of the returned promise if preferred over callback.

File: app.json [hunk 13]
```diff
@@ -3076,8 +3118,8 @@
               "value": "0",
               "pattern": "^0$|^[1-6];([1-6];)?([1-6];)?([1-6];)?[1-6];?",
               "hint": {
-                "en": "Separate button numbers by a semicolon ( ; ), use 2 to 5 buttons. No duplicate sequences are allowed. 0 = sequence disabled. Buttons are numbered from left to right, from top to bottom (1 - 6).",
-                "nl": "Scheid knop nummers door een puntkomma ( ; ), gebruik 2 tot 5 knoppen. 2x dezelfde reeks is niet toegestaan. 0 = reeks uitgeschakeld. Knoppen zijn genummerd van links naar rechts, van boven naar beneden (1 - 6)."
+                "en": "This parameter 1 sets a button sequence in the device. Separate button numbers by a semicolon ( ; ), use 2 to 5 buttons. No duplicate sequences are allowed. 0 = sequence disabled. Buttons are numbered from left to right, from top to bottom (1 - 6).",
```
Comment (RobinBol) [inline app.json]: (1)*

File: app.json [hunk 14]
```diff
@@ -3057,14 +3099,14 @@
...
```
Comment (RobinBol) [inline app.json]: (1)*

File: app.json [hunk 15]
```diff
@@ -2966,10 +2968,50 @@
...
+                "index": 1,
+                "size": 2,
+                "signed": false
+              },
+              "id": "sequence_lock",
+              "type": "text",
+              "value": "0",
+              "pattern": "^0$|^[1-6];([1-6];)?([1-6];)?([1-6];)?[1-6];?",
+              "hint": {
+                "en": "This parameter 1 sets a button sequence in the device. Separate button numbers by a semicolon ( ; ), use 2 to 5 buttons. No duplicate sequences are allowed. 0 = sequence disabled. Buttons are numbered from left to right, from top to bottom (1 - 6).",
```
Comment (RobinBol) [inline app.json]: (1)* (multiple times)

File: drivers/FGKF-601/device.js (hunk 11 above)
Comment (MatthewWaanders) [reply drivers/FGKF-601/device.js]: Wasn't, now fixed

File: drivers/FGKF-601/device.js (hunk 12 above)
Comment (MatthewWaanders) [reply drivers/FGKF-601/device.js]: Fixed

File: app.json (hunk 13 above)
Comment (MatthewWaanders) [reply app.json]: Fixed in setting extended, applies to all

File: drivers/FGKF-601/device.js [hunk 16]
```diff
@@ -64,38 +65,36 @@ class FibaroKeyfob extends ZwaveDevice {
     }
 
     _enableLockMode(newValueObj) {
-	    if ( (typeof newValueObj.lock_timeout === 'number' &&
-                newValueObj.lock_timeout > 0 &&
-                this.getSetting('sequence_lock')) ||
+	    if ((typeof newValueObj.lock_timeout === 'number' &&
+            newValueObj.lock_timeout > 0 &&
```
Comment (RobinBol) [inline drivers/FGKF-601/device.js]: Weird indentation, I think an extra tab is preferred when continuing an if statement on a new line.

File: drivers/FGKF-601/device.js [hunk 17]
```diff
@@ -64,38 +65,36 @@ class FibaroKeyfob extends ZwaveDevice {
...
+            this.getSetting('sequence_lock')) ||
             (typeof newValueObj.sequence_lock === 'string' &&
-                this.getSetting('lock_timeout') > 0)) {
-	                this.node.CommandClass.COMMAND_CLASS_PROTECTION.PROTECTION_SET({
+            this.getSetting('lock_timeout') > 0)) {
+                try {
+                    this.node.CommandClass.COMMAND_CLASS_PROTECTION.PROTECTION_SET({
```
Comment (RobinBol) [inline drivers/FGKF-601/device.js]: This is not awaited?

File: drivers/FGKF-601/device.js [hunk 18]
```diff
@@ -64,38 +65,36 @@ class FibaroKeyfob extends ZwaveDevice {
...
-                if (err) this.log(err);
-                if (result === 'TRANSMIT_COMPLETE_OK') {
-                    this.log('Succesfully unset protection command');
+                } catch (err) {
+                    this.error(err);
                 }
-            });
+        } else {
+	        try {
+                this.node.CommandClass.COMMAND_CLASS_PROTECTION.PROTECTION_SET({
```
Comment (RobinBol) [inline drivers/FGKF-601/device.js]: This is not awaited?

File: drivers/FGKF-601/device.js [hunk 19]
```diff
@@ -13,20 +13,30 @@ class FibaroKeyfob extends ZwaveDevice {
         this._sequenceFlowTrigger = new Homey.FlowCardTriggerDevice('FGKF-601-sequence').registerRunListener(this._sequenceRunListener).register();
 
         // Parsing of sequences before sending to Keyfob
-        this.registerSetting('lock_timeout', async (newValue) => {
+        this.registerSetting('lock_timeout', async (oldValue, newValue) => {
```
Comment (RobinBol) [inline drivers/FGKF-601/device.js]: Where does the `oldValue` come from?

File: drivers/FGKF-601/device.js [hunk 20]
```diff
@@ -13,20 +13,30 @@ class FibaroKeyfob extends ZwaveDevice {
...
             try {
+                let requestTimeout = setTimeout(() => {
+                    Promise.reject('Device Sleeping');
+                }, 500);
```
Comment (RobinBol) [inline drivers/FGKF-601/device.js]: 0.5 seconds is a bit to quick I think;)

File: drivers/FGKF-601/device.js [hunk 21]
```diff
@@ -13,20 +13,30 @@ class FibaroKeyfob extends ZwaveDevice {
         this._sequenceFlowTrigger = new Homey.FlowCardTriggerDevice('FGKF-601-sequence').registerRunListener(this._sequenceRunListener).register();
 
         // Parsing of sequences before sending to Keyfob
-        this.registerSetting('lock_timeout', async (newValue) => {
+        this.registerSetting('lock_timeout', async (oldValue, newValue) => {
             try {
+                let requestTimeout = setTimeout(() => {
+                    Promise.reject('Device Sleeping');
```
Comment (RobinBol) [inline drivers/FGKF-601/device.js]: `TRANSMIT_QUEUED` is bit more descriptive I think.

Comment (RobinBol) [reply drivers/FGKF-601/device.js]: Oh nevermind, I thought you resolved instead of rejected, `device_asleep` is probably a more usable string.

**Pull request merged**